File: dashboard-ts/src/shared_copy/utils/logger_config.py
"""
統一日誌配置模組

提供標準化的 JSON 格式日誌配置
- 日誌輪轉：每日或達到 500MB
- 保留期限：30 天
- 格式：JSON
"""
import sys
from pathlib import Path
from loguru import logger


def setup_logger(
    service_name: str,
    log_dir: str = "logs",
    level: str = "INFO",
    rotation: str = "500 MB",
    retention: str = "30 days",
    json_format: bool = True
):
    """設定統一日誌配置

    Args:
        service_name: 服務名稱（如：collector, analyzer）
        log_dir: 日誌目錄
        level: 日誌等級
        rotation: 輪轉條件（"500 MB" 或 "1 day"）
        retention: 保留期限（"30 days"）
        json_format: 是否使用 JSON 格式
    """
    # 移除預設 handler
    logger.remove()

    # 確保日誌目錄存在
    log_path = Path(log_dir)
    log_path.mkdir(parents=True, exist_ok=True)

    # JSON 格式配置
    if json_format:
        log_format = (
            "{{\"timestamp\": \"{time:YYYY-MM-DD HH:mm:ss.SSS}\", "
            "\"level\": \"{level}\", "
            "\"service\": \"" + service_name + "\", "
            "\"module\": \"{name}\", "
            "\"function\": \"{function}\", "
            "\"line\": {line}, "
            "\"message\": \"{message}\"}}\n"
        )
    else:
        log_format = (
            "<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | "
            "<level>{level: <8}</level> | "
            "<cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> | "
            "<level>{message}</level>"
        )

    # 添加控制台 handler（人類可讀格式）
    logger.add(
        sys.stderr,
        format=log_format if not json_format else log_format,
        level=level,
        colorize=not json_format,
    )

    # 添加檔案 handler（JSON 格式，有輪轉）
    logger.add(
        log_path / f"{service_name}.log",
        format=log_format,
        level=level,
        rotation=rotation,
        retention=retention,
        compression="zip",  # 壓縮舊日誌
        enqueue=True,  # 非同步寫入
        backtrace=True,  # 記錄完整 traceback
        diagnose=True,  # 記錄變數值
    )

    # 添加錯誤專用檔案 handler
    logger.add(
        log_path / f"{service_name}.error.log",
        format=log_format,
        level="ERROR",
        rotation=rotation,
        retention=retention,
        compression="zip",
        enqueue=True,
        backtrace=True,
        diagnose=True,
    )

    logger.info(f"{service_name} 日誌系統初始化完成")
    logger.info(f"日誌目錄：{log_path.absolute()}")
    logger.info(f"日誌等級：{level}")
    logger.info(f"輪轉條件：{rotation}")
    logger.info(f"保留期限：{retention}")
    logger.info(f"JSON 格式：{json_format}")

    return logger


# 預設配置
def get_default_logger(service_name: str):
    """獲取預設配置的 logger

    Args:
        service_name: 服務名稱

    Returns:
        配置好的 logger
    """
    import os

    return setup_logger(
        service_name=service_name,
        log_dir=os.getenv("LOG_DIR", "logs"),
        level=os.getenv("LOG_LEVEL", "INFO"),
        rotation=os.getenv("LOG_ROTATION", "500 MB"),
        retention=os.getenv("LOG_RETENTION", "30 days"),
        json_format=os.getenv("LOG_JSON", "true").lower() == "true"
    )

File: shared/utils/logger_config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupLogger = setupLogger;
exports.getDefaultLogger = getDefaultLogger;
/**
 * 統一日誌配置模組 (Node.js)
 *
 * 提供標準化的 JSON 格式日誌配置
 * - 日誌輪轉：每日或達到 500MB
 * - 保留期限：30 天
 * - 格式：JSON
 */
const winston_1 = __importDefault(require("winston"));
const winston_daily_rotate_file_1 = __importDefault(require("winston-daily-rotate-file"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
/**
 * 設定統一日誌配置
 *
 * @param config - 日誌配置
 * @returns 配置好的 winston logger
 */
function setupLogger(config) {
    const { serviceName, logDir = 'logs', level = 'info', maxSize = '500m', maxFiles = '30d', jsonFormat = true, } = config;
    // 確保日誌目錄存在
    if (!fs_1.default.existsSync(logDir)) {
        fs_1.default.mkdirSync(logDir, { recursive: true });
    }
    // 自定義格式
    const customFormat = winston_1.default.format.combine(winston_1.default.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }), winston_1.default.format.errors({ stack: true }), winston_1.default.format.metadata(), jsonFormat
        ? winston_1.default.format.json()
        : winston_1.default.format.printf(({ timestamp, level, message, service, ...meta }) => {
            const metaStr = Object.keys(meta).length ? JSON.stringify(meta, null, 2) : '';
            return `${timestamp} [${level.toUpperCase()}] [${service}] ${message} ${metaStr}`;
        }));
    // 一般日誌輪轉配置
    const fileRotateTransport = new winston_daily_rotate_file_1.default({
        filename: path_1.default.join(logDir, `${serviceName}-%DATE%.log`),
        datePattern: 'YYYY-MM-DD',
        maxSize,
        maxFiles,
        level,
        format: customFormat,
        zippedArchive: true, // 壓縮舊日誌
    });
    // 錯誤日誌輪轉配置
    const errorRotateTransport = new winston_daily_rotate_file_1.default({
        filename: path_1.default.join(logDir, `${serviceName}.error-%DATE%.log`),
        datePattern: 'YYYY-MM-DD',
        maxSize,
        maxFiles,
        level: 'error',
        format: customFormat,
        zippedArchive: true,
    });
    // 控制台輸出
    const consoleTransport = new winston_1.default.transports.Console({
        level,
        format: winston_1.default.format.combine(winston_1.default.format.colorize(), winston_1.default.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }), winston_1.default.format.printf(({ timestamp, level, message, service }) => {
            return `${timestamp} ${level} [${service}] ${message}`;
        })),
    });
    // 建立 logger
    const logger = winston_1.default.createLogger({
        level,
        defaultMeta: { service: serviceName },
        transports: [consoleTransport, fileRotateTransport, errorRotateTransport],
        exceptionHandlers: [
            new winston_daily_rotate_file_1.default({
                filename: path_1.default.join(logDir, `${serviceName}.exceptions-%DATE%.log`),
                datePattern: 'YYYY-MM-DD',
                maxSize,
                maxFiles,
                zippedArchive: true,
            }),
        ],
        rejectionHandlers: [
            new winston_daily_rotate_file_1.default({
                filename: path_1.default.join(logDir, `${serviceName}.rejections-%DATE%.log`),
                datePattern: 'YYYY-MM-DD',
                maxSize,
                maxFiles,
                zippedArchive: true,
            }),
        ],
    });
    logger.info(`${serviceName} 日誌系統初始化完成`, {
        logDir: path_1.default.resolve(logDir),
        level,
        maxSize,
        maxFiles,
        jsonFormat,
    });
    return logger;
}
/**
 * 獲取預設配置的 logger
 *
 * @param serviceName - 服務名稱
 * @returns 配置好的 logger
 */
function getDefaultLogger(serviceName) {
    return setupLogger({
        serviceName,
        logDir: process.env.LOG_DIR || 'logs',
        level: process.env.LOG_LEVEL || 'info',
        maxSize: process.env.LOG_MAX_SIZE || '500m',
        maxFiles: process.env.LOG_MAX_FILES || '30d',
        jsonFormat: process.env.LOG_JSON !== 'false',
    });
}
